server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java

```diff
@@ -61,25 +64,33 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
         assert tasks.stream().allMatch(t -> t.batchingKey == firstTask.batchingKey) :
             "tasks submitted in a batch should share the same batching key: " + tasks;
         // convert to an identity map to check for dups based on task identity
-        final Map<Object, BatchedTask> tasksIdentity = tasks.stream().collect(Collectors.toMap(
-            BatchedTask::getTask,
+        final Map<IdentityWrapper, BatchedTask> toAdd = tasks.stream().collect(Collectors.toMap(
+            t -> new IdentityWrapper(t.getTask()),
             Function.identity(),
             (a, b) -> { throw new IllegalStateException("cannot add duplicate task: " + a); },
-            IdentityHashMap::new));
-
-        synchronized (tasksPerBatchingKey) {
-            LinkedHashSet<BatchedTask> existingTasks = tasksPerBatchingKey.computeIfAbsent(firstTask.batchingKey,
-                k -> new LinkedHashSet<>(tasks.size()));
-            for (BatchedTask existing : existingTasks) {
-                // check that there won't be two tasks with the same identity for the same batching key
-                BatchedTask duplicateTask = tasksIdentity.get(existing.getTask());
-                if (duplicateTask != null) {
-                    throw new IllegalStateException("task [" + duplicateTask.describeTasks(
-                        Collections.singletonList(existing)) + "] with source [" + duplicateTask.source + "] is already queued");
+            LinkedHashMap::new));
```

Comment:
why LinkedHashMap? Do you care about insertion order?

Reply:
I have been trying to understand that and I am not 100% sure that we should or should not care.
The insertion order was maintained in the original version with the help of `LinkedHashSet` though and I was not brave enough to relax the semantics.